### ISSUE TITLE
Dynamic coding for GridPROTEUS

### ISSUE THREAD

**Issue by nichollsh** [FormingWorlds/PROTEUS#162]
Currently, a large amount of the configuration of GridPROTEUS.py is hardcoded in the `if __name__` statement. This would ideally be configured in another manner, such as by a configuration file or with a CLI.

**Comment by stefsmeets** [comment]
I would suggest to hold off major work on this until we refactored the proteus (single) run code. My hope is to fold `GridProteus` into there as well eventually.

**Comment by EmmaPostolec** [comment]
Hi, I edited tools/grid_proteus.py with new parameters that I wanted to test for the escape (Pxuv, efficiency and semimajoraxis) and I ran into the following error:

````
[2024-11-13 11:25:32] Sleeping...
[2024-11-13 11:25:32]     7 
[2024-11-13 11:25:33]     6 
[2024-11-13 11:25:34]     5 
[2024-11-13 11:25:35]     4 
[2024-11-13 11:25:36]     3 
[2024-11-13 11:25:37]     2 
[2024-11-13 11:25:38]     1 
[2024-11-13 11:25:39]  
[2024-11-13 11:25:39] Writing config files
[2024-11-13 11:25:45] Starting process manager (75 grid points, 40 threads)
[2024-11-13 11:25:45]  75 queued (100.0%),   0 running (  0.0%),   0 completed (  0.0%)
[2024-11-13 11:25:45] Uncaught exception
Traceback (most recent call last):
  File "/Users/emmapostolec/Documents/PHD/SCIENCE/CODES/PROTEUS/tools/grid_proteus.py", line 492, in <module>
    pg.run(40, test_run=False)
  File "/Users/emmapostolec/Documents/PHD/SCIENCE/CODES/PROTEUS/tools/grid_proteus.py", line 405, in run
    threads[i].start()
  File "/Users/emmapostolec/miniconda3/envs/proteus/lib/python3.12/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/Users/emmapostolec/miniconda3/envs/proteus/lib/python3.12/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/emmapostolec/miniconda3/envs/proteus/lib/python3.12/multiprocessing/context.py", line 289, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/emmapostolec/miniconda3/envs/proteus/lib/python3.12/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/emmapostolec/miniconda3/envs/proteus/lib/python3.12/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/emmapostolec/miniconda3/envs/proteus/lib/python3.12/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/emmapostolec/miniconda3/envs/proteus/lib/python3.12/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'Grid.run.<locals>._thread_target'
````
@nichollsh suggested having ````multiprocessing.set_start_method('fork')```` on line 25, and with this change it seems to work on my computer now! Should I update grid_proteus and open a new pull request to merge into main ? Let me know what you think is the best. @nichollsh @timlichtenberg @lsoucasse @stefsmeets

**Comment by stefsmeets** [comment]
Do you know when the error was introduced?

**Comment by nichollsh** [comment]
This might have been here since the script was created. I have not found this error on my system at any point in the past. 

This article (with the fix mentioned above) suggest that it depends on the platform in some strange way. https://medium.com/devopss-hole/python-multiprocessing-pickle-issue-e2d35ccf96a9

**Comment by stefsmeets** [comment]
You could argue whether fork vs. spawn is better, but note that fork is not available on Windows, buggy on macs, and discouraged because forking has lots of undefined behaviour. This can lead to very hard to reproduce bugs.

Looking at the code, my take is that this code should use threading module instead of multiprocessing (e.g. via a `Threadpool`). The code already calls a subprocess for the actual work, so that means that every task essentiall creates 2 processes, which is very inefficient.

**Comment by nichollsh** [comment]
There's definitely a lot of room for improvement in this script. Using threadpools makes more sense than the current bespoke formulation for queueing. Ideally, it would also be an option to run the grid across multiple nodes using the Python SLURM library.

This is not particularly high-priority at the moment though - as long as the current script works for everyone.

**Comment by stefsmeets** [comment]
I agree with your assesment that it is not very high priority. SLURM or subprocess is not the major development bottleneck here imo, if the rest of the code is well factored.

**Comment by timlichtenberg** [comment]
Coming back to this a bit, because I believe dynamic coding for grid runs is quite essential to the functioning of the code. How could this be implemented with the use of the current config files?

For example, instead of
`mass        = 1.0       # M_earth`
could the user set:
`mass        = [1.0, 1.5, 2.0]       # M_earth`
and the code automatically detects this to be a grid run, and accordingly launches multiple simulations in parallel?

If not, what are the alternatives?

**Comment by nichollsh** [comment]
I like this idea a lot. It's intuitive and doesn't require the user to learn a new configuration file format.

**Comment by timlichtenberg** [comment]
This is a good idea and will simplify the implementation. I want to shortly connect this to #204. Assuming we would [pymultinest](https://github.com/JohannesBuchner/PyMultiNest) for this implementation, the call function to run a PROTEUs grid will be important to also pass to an external module. Alternatively, pymultinest can become  a PROTEUS module, and the config file would get an addition "retrieval" section of parameters that can be retrieved. Once this flag is passed (for example, via `proteus retrieval -c file.toml`), the user could specify both the priors of all model parameters and the data (i.e., mass, radius, age, spectrum, for example). This latter option would homogenise the PROTEUS call sequence for both forward and inverse modelling approaches.

**Comment by stefsmeets** [comment]
I like the idea as well, but overloading the config object in this way will be a hugely complex task. There are libraries that can help with this sort of work (e.g. for use in uncertainty quantification studies).

**Comment by nichollsh** [comment]
I can see how this might get complicated under the current formulation. Which libraries did you have in mind?

**Comment by timlichtenberg** [comment]
Next steps to prepare PROTEUS for an inverse modelling approach:

- [ ] Make PROTEUS pip installable
- [ ] Write a connector (python API) for connecting PROTEUS input & output to grid/UQ model
- [ ] Investigate which source model to use (see some suggestions above)
- [ ] Test the model connection with one or two options

Further down the line:
- [ ] Apply for further grants :-)
- [ ] Make PROTEUS more efficient to enable larger grids and UQ

**Comment by nichollsh** [comment]
https://exoplanetarchive.ipac.caltech.edu/cgi-bin/atmospheres/nph-firefly
